helper/gate/gate.go

```diff
@@ -0,0 +1,87 @@
+// Package gate implements a simple on/off latch or gate: it blocks waiters
```

Comment:
TR.Update runs concurrently with Client.runAllocs, both of which may open the gate. This means I couldn't simply close an unsynchronized chan to unblock waiting TRs. I decided to encapsulate the concurrent unblockers in this gate package as it also provides a concise API (IsClosed) for determining whether a TR is about to block and logging beforehand.